File: DataStructures/LinkedList/python/LinkedList.py
import sys

class Node:
	def __init__(self, data, nexxt = None):
		self.data = data
		self.next = nexxt

class LList:
	def __init__(self):
		self.head = None

	def insert(self, data):
		newnode = Node(data)
		if self.head is None:
			self.head = newnode
		else:
			temp = self.head
			while temp.next is not None:
				temp = temp.next
			temp.next = newnode 

	def insert_at_head(self, data):
		newnode = Node(data)
		if self.head is None:
			self.head = newnode
		else:
			newnode.next = self.head 
			self.head = newnode

	def search(self, val):
		temp = self.head
		while(temp):
			if(temp.data == val):
				return True
			temp = temp.next
		return False

	def remove(self, val):
		temp = self.head
		prev = None
		while(temp):
			if(temp.data == val):
				break
			prev = temp
			temp = temp.next
		if(temp):
			if(prev is None):
				self.head = temp.next
			else:
				prev.next = temp.next
			temp = None
		self.display()

	def reverse(self, head):
		if head.next is None:
			self.head = head
			return
		self.reverse(head.next)
		head.next.next = head
		head.next = None

	#private method should start with two underscores '__example()'
	def __split(self, head):    
		if head is None or head.next is None:
			return head, None
		slowptr = head
		fastptr = head.next
		while(fastptr):
			fastptr = fastptr.next
			if(fastptr):
				slowptr = slowptr.next
				fastptr = fastptr.next
		fHalf = head
		sHalf = slowptr.next
		slowptr.next = None
		return fHalf, sHalf

	def __merge(self, a, b):
		if a is None:
			return b
		if b is None:
			return a
		temp = None
		if a.data <= b.data:
			temp = a
			temp.next = self.__merge(a.next, b)
		else:
			temp = b
			temp.next = self.__merge(a, b.next)
		return temp		

	def mergeSort(self, head):
		if head is None or head.next is None:
			return head
		fHalf, sHalf = self.__split(head)
		fHalf = self.mergeSort(fHalf)
		sHalf = self.mergeSort(sHalf)
		return self.__merge(fHalf, sHalf)

	def display(self):
		temp = self.head
		while(temp):
			print(temp.data,end = " --> ")
			temp = temp.next
		print("null")

	def displayReverse(self, head):
		if head is None:
			return
		self.displayReverse(head.next)
		print(head.data, end = " ")

if __name__ == '__main__':

	sys.stdin = open('../input.txt', 'r')
	sys.stdout = open('../output.txt', 'w')

	n = sys.stdin.readline();
	a = sys.stdin.readline().split(" ")[:-1] # [:-1] to eliminate reading of newline
	a = [int(i) for i in a]

	slist = LList()

	#Insertion
	print("Insertion at tail",end = "\n")
	for i in a:
		slist.insert(i)
	slist.display()

	slist2 = LList()
	print("Insertion at head",end = "\n")
	for i in a:
		slist2.insert_at_head(i)
	slist2.display()

	#search
	sval = int(sys.stdin.readline())
	isAvailable = slist.search(sval)
	if(isAvailable):
		print(sval," is found.")
	else:
		print(sval," is not found.")

	#delete
	dval = int(sys.stdin.readline())
	print("After deletion of node ", dval,end = "\n")
	slist.remove(dval)

	#reverse
	print("Reversing List", end = "\n")
	slist.reverse(slist.head)
	slist.display()
	
	# traverse reverse
	print("Traversing reverse", end = "\n")
	slist.displayReverse(slist.head)

	# merge sort
	print("Sorted List", end = "\n")
	slist.head = slist.mergeSort(slist.head)
	slist.display()

File: DataStructures/LinkedList/c/LinkedList.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

struct node{
	int data;
	struct node* next;
};

struct node* insert_at_tail(struct node* head, int data){
	struct node* newnode = (struct node*)malloc(sizeof(struct node));
	if(newnode == NULL){
		printf("Memory is not allocated");
		exit(1); 
	}
	newnode -> data = data;
	newnode -> next = NULL;
	if(head == NULL){
		head = newnode;
		return head;
	}
	struct node* temp = head;
	while(temp->next != NULL){
		temp = temp -> next;
	}
	temp -> next = newnode;
	return head;
}

struct node* insert_at_head(struct node* head, int data){
	struct node* newnode = (struct node*)malloc(sizeof(struct node));
	if(newnode == NULL){
		printf("Memory is not allocated");
		exit(1); 
	}
	newnode -> data = data;
	newnode -> next = head;
	head = newnode;
	return head;
}

bool search(struct node* head, int val){
	struct node* temp = head;
	while(temp != NULL){
		if(temp -> data == val){
			return true;
		}
		temp = temp -> next;
	}
	return false;
}

struct node* _remove(struct node* head, int val){
	struct node* temp = head;
	struct node* prev = NULL;
	while(temp != NULL){
		if(temp -> data == val){
			break;
		}
		prev = temp;
		temp = temp -> next;
	}
	if(temp != NULL){
		if(prev == NULL){
			struct node* delnode = temp;
			head = temp -> next;
			free(delnode);
		}
		else{
			struct node* delnode = temp;
			prev -> next = temp -> next;
			free(delnode);
		}
	}
	return head;
}

struct node* reverseList(struct node* head){
	if(head -> next == NULL){
		return head;
	}
	struct node* h = reverseList(head -> next);
	head -> next -> next = head;
	head -> next = NULL;
	return h;
}

void split(struct node* head, struct node** fh, struct node** sh){
	if(head == NULL || head->next == NULL){
		*fh = head;
		*sh = NULL;
		return; 
	}
	struct node* slowptr = head;
	struct node* fastptr = head -> next;
	while(fastptr != NULL){
		fastptr = fastptr -> next;
		if(fastptr != NULL){
			slowptr = slowptr -> next;
			fastptr = fastptr -> next;
		}
	}
	*fh = head;
	*sh = slowptr -> next;
	slowptr -> next = NULL;
	return;

}
struct node* merge(struct node* fh, struct node* sh){
	if(fh == NULL){
		return sh;
	}
	if(sh == NULL){
		return fh;
	}
	struct node* temp = NULL;
	if(fh->data <= sh->data){
		temp = fh;
		temp -> next = merge(fh->next, sh);
	}
	else{
		temp = sh;
		temp -> next = merge(fh, sh->next);
	}
	return temp;
}
void mergeSort(struct node **head){
	if((*head) == NULL || (*head)->next == NULL)
		return;
	struct node *fHalf, *sHalf;
	split(*head, &fHalf, &sHalf);
	mergeSort(&fHalf);
	mergeSort(&sHalf);
	*head = merge(fHalf, sHalf);
}

void display(struct node* head){
	while(head != NULL){
		printf("%d --> ", head->data);
		head = head -> next;
	}
	printf("null\n");
}

void displayReverse(struct node* head){
	if(head == NULL)
		return;
	displayReverse(head -> next);
	printf("%d ", head -> data);
}

int main(){

	freopen("../input.txt","r",stdin);
	freopen("../output.txt","w",stdout);

	int n,i;
	scanf("%d",&n);
	int a[n];
	for(i=0; i<n; i++)
		scanf("%d",&a[i]);

	//Insertion 
	printf("insertion at tail\n");
	struct node* head = NULL;
	for(i=0; i<n; i++){
		head = insert_at_tail(head,a[i]);
	}
	display(head); // Traversing list
 
	printf("insertion at head\n");
	head = NULL;
	for(i=0; i<n; i++){
		head = insert_at_head(head,a[i]);
	}
	display(head);

	//Search
	int searchVal;
	scanf("%d", &searchVal);
	bool isAvailable = search(head,searchVal);
	if(isAvailable)
		printf("%d is found.\n", searchVal);
	else
		printf("%d is not found.\n", searchVal);

	//Delete
	int delVal;
	scanf("%d", &delVal);
	head = _remove(head, delVal);
	printf("After deletion of node %d\n", delVal);
	display(head);

	//Reverse
	head = reverseList(head);
	printf("Reversing list\n");
	display(head);

	//Traverse reverse
	displayReverse(head);

	//Merge Sort
	mergeSort(&head);
	printf("Sorted List\n");
	display(head);
}